hermes/probe/create_file.go

```diff
@@ -6,109 +6,119 @@ import (
+	i    int // currently reading this byte
 	rand *rand.Rand
 }
 
-func (r *randomFileReader) Read(b []byte) (n int, err error) {
-	if r.i >= r.size {
+func (r *randomHermesFileReader) readDone() bool {
+	return r.i >= r.size
+}
+
+func (r *randomHermesFileReader) bytesLeft() int {
+	return (r.size - r.i)
+}
+
+func (r *randomHermesFileReader) bufferTooLong(buffer []byte) bool {
+	return len(buffer) > r.bytesLeft()
+}
+
+func (r *randomHermesFileReader) Read(buff []byte) (n int, err error) {
+	if r.readDone() {
 		return 0, io.EOF
 	}
-	if len(b) > r.size-r.i {
-		n, err = r.rand.Read(b[:r.size-r.i])
-	} else {
-		n, err = r.rand.Read(b)
+	b := buff
+	if r.bufferTooLong(b){
+		b = buff[:r.bytesLeft()]
+	}
+	n, err = r.rand.Read(b)
+	if(err!=nil){
+		return 0, err // 0 as nil casn't be returned as a type int argument
 	}
 	r.i += n
 	return
 }
 
-func (f *RandomFile) generateChecksum() (string, error) {
+// Warning: NewReader is not thread safe
+func (f *RandomHermesFile) NewReader() *randomHermesFileReader {
+	return &randomHermesFileReader{size: f.Size, rand: rand.New(rand.NewSource(f.ID))} // ID will serve as a Seed and i will be set to 0 automatically
+}
+
+func (f *RandomHermesFile) CheckSum() ([]byte, error) {
 	r := f.NewReader()
 	h := sha1.New()
-	buff := make([]byte, 8)
-	for {
-		n, err := r.Read(buff)
-		if err == io.EOF {
-			break
-		}
-		if _, err = io.Copy(h, r); err != nil {
-			return nil, fmt.Errorf("io.Copy: %v", err)
-		}
-
+	if _, err := io.Copy(h, r); err != nil {
+		return nil, fmt.Errorf("io.Copy: %v", err)
 	}
-
 	// returns checksum in hex notation
-	return fmt.Sprintf("%x", h.Sum(nil)), nil
+	return h.Sum(nil), nil
 }
 
-func (f *RandomFile) generateFileName(fileID int) (string, error) {
-	checksum, err := f.generateChecksum()
+func (f *RandomHermesFile) FileName() (string, error) {
+	if f.ID < begin || f.ID > end {
+		return "", fmt.Errorf("The file ID provided %v wasn't in the required range [1,50]", f.ID)
+	}
+	if f.Size > maxFileSize {
+		return "", fmt.Errorf("The file size provided %v bytes exceeded the limit %v bytes", f.Size, maxFileSize)
+	}
+	if f.Size <= 0 {
+		return "", fmt.Errorf("The file size provided %v is not a positive number as required", f.Size)
+	}
+	checksum, err := f.CheckSum()
 	if err != nil {
-		return nil, err
+		return "", err
 	}
-	return fmt.Sprintf("Hermes_%02d_%v", fileID, checksum), nil
+	return fmt.Sprintf("Hermes_%02d_%v", f.ID, fmt.Sprintf("%x", checksum)), nil
 }
 
-func (p *MonitorProbe) CreateFile(ctx context.Context, target string, bucketName string, fileID int, fileSize int) error {
-	if fileID < 1 || fileID > 50 {
-		return fmt.Errorf("At %v the file ID provided wasn't in the required range [0,50]", time.Now())
-	}
-	f := RandomFile{Seed: fileID, Size: fileSize}
-	fileName, err := f.generateFileName(fileID)
+func (p *MonitorProbe) CreateFile(ctx context.Context, target string, bucketName string, fileID int64, fileSize int, client stiface.Client) error {
```

Comment:
- [ ] Would it make more sense to make these a methods of Targets instead? So that  a MonitorProbe (who has a field `targets` of  type `[]*Target` can make calls such as `t.CreateFile()`. Plus you don't appear to use "p" any more
- [  ] `target` shouldn't be a string now that there is a `Target` struct in monitor.go.
- [ ] Make client a pointer instead
- [ ] If you need the logger of a MonitorProbe, you can pass that as a pointer within the arguments.

Reply:
Oh, if you keep `p`, rename it to `mp` just to keep it consistent with the rest.

Reply:
p removed, CreateFile is now a function and it takes in Targets as an argument as they will be defined in a separate file and maybe even package
Client was changed to a pointer but now will be changed back, because as @leczb pointed out interfaces are stored as a tuple (with first parameter pointer and the second one can be nil so they are in themselves pointers)